host/versioning_test.go

```diff
@@ -0,0 +1,209 @@
+		Namespace: s.namespace,
+		TaskQueue: tq,
+	})
+	s.NoError(err)
+	s.NotNil(res2)
+	s.Equal(res2.CurrentDefault.GetVersion().GetWorkerBuildId(), "foo")
+}
+
+func (s *integrationSuite) TestSeriesOfUpdates() {
+	ctx := NewContext()
+	tq := "integration-versioning-series"
+	s.prepareQueue(ctx, tq)
+
+	for i := 0; i < 10; i++ {
+		res, err := s.engine.UpdateWorkerBuildIdOrdering(ctx, &workflowservice.UpdateWorkerBuildIdOrderingRequest{
+			Namespace:          s.namespace,
+			TaskQueue:          tq,
+			VersionId:          &taskqueuepb.VersionId{WorkerBuildId: fmt.Sprintf("foo-%d", i)},
+			PreviousCompatible: nil,
+			BecomeDefault:      true,
+		})
+		s.NoError(err)
+		s.NotNil(res)
+	}
+	res, err := s.engine.UpdateWorkerBuildIdOrdering(ctx, &workflowservice.UpdateWorkerBuildIdOrderingRequest{
+		Namespace:          s.namespace,
+		TaskQueue:          tq,
+		VersionId:          &taskqueuepb.VersionId{WorkerBuildId: "foo-2.1"},
+		PreviousCompatible: &taskqueuepb.VersionId{WorkerBuildId: "foo-2"},
+		BecomeDefault:      false,
+	})
+	s.NoError(err)
+	s.NotNil(res)
+
+	res2, err := s.engine.GetWorkerBuildIdOrdering(ctx, &workflowservice.GetWorkerBuildIdOrderingRequest{
+		Namespace: s.namespace,
+		TaskQueue: tq,
+	})
+	s.NoError(err)
+	s.NotNil(res2)
+	s.Equal(res2.CurrentDefault.GetVersion().GetWorkerBuildId(), "foo-9")
+	s.Equal(len(res2.CompatibleLeaves), 1)
+	s.Equal(res2.CompatibleLeaves[0].GetVersion().GetWorkerBuildId(), "foo-2.1")
+}
+
+func (s *integrationSuite) TestLinkToNonexistentCompatibleVersionReturnsNotFound() {
+	ctx := NewContext()
+	tq := "integration-versioning-compat-not-found"
+	s.prepareQueue(ctx, tq)
+
+	res, err := s.engine.UpdateWorkerBuildIdOrdering(ctx, &workflowservice.UpdateWorkerBuildIdOrderingRequest{
+		Namespace:          s.namespace,
+		TaskQueue:          tq,
+		VersionId:          &taskqueuepb.VersionId{WorkerBuildId: "foo"},
+		PreviousCompatible: &taskqueuepb.VersionId{WorkerBuildId: "i don't exist yo"},
+	})
+	s.Error(err)
+	s.Nil(res)
+	s.IsType(&serviceerror.NotFound{}, err)
+}
+
+func (s *integrationSuite) TestVersioningStateNotDestroyedByOtherUpdates() {
+	ctx := NewContext()
+	tq := "integration-versioning-not-destroyed"
+	s.prepareQueue(ctx, tq)
+
+	res, err := s.engine.UpdateWorkerBuildIdOrdering(ctx, &workflowservice.UpdateWorkerBuildIdOrderingRequest{
+		Namespace:          s.namespace,
+		TaskQueue:          tq,
+		VersionId:          &taskqueuepb.VersionId{WorkerBuildId: "foo"},
+		PreviousCompatible: nil,
+		BecomeDefault:      true,
+	})
+	s.NoError(err)
+	s.NotNil(res)
+
+	isFirst := true
+	wtHandler := func(execution *commonpb.WorkflowExecution, wt *commonpb.WorkflowType,
+		previousStartedEventID, startedEventID int64, history *historypb.History) ([]*commandpb.Command, error) {
+		// TODO: This timer is long to ensure the 1-minute lease-renewal on the task queue happens, to verify that
+		//   doesn't blow up data. There must be a faster way to do that.
```

Comment:
K, I've done something similar to but not exactly this, as I need the config to be set properly before anything starts up

Reply:
yeah, I was hoping it could be per-test, but thinking a little more, that seems impossible since everything is running by the time each test starts. they are theoretically _dynamic_ configs, so they could be changed after startup, but many of them aren't actually dynamic. so per-suite might be as good as we can reasonably do